crates/hotshot/src/tasks/mod.rs

```diff
@@ -55,165 +44,93 @@ pub enum GlobalEvent {
+        loop {
+            let msgs = match network.recv_msgs(TransmitType::Broadcast).await {
+                Ok(msgs) => Messages(msgs),
+                Err(err) => {
+                    error!("failed to receive broadcast messages: {err}");
 
-                if msgs.0.is_empty() {
-                    async_sleep(Duration::from_millis(100)).await;
-                } else {
-                    break msgs;
+                    // return zero messages so we sleep and try again
+                    Messages(vec![])
                 }
+            };
+            if msgs.0.is_empty() {
+                async_sleep(Duration::from_millis(100)).await;
+            } else {
+                state.handle_messages(msgs.0).await;
             }
-        };
-        Some(boxed_sync(closure))
-    }));
-    let net = channel.clone();
-    let direct_stream = GeneratedStream::<Messages<TYPES>>::new(Arc::new(move || {
-        let network = net.clone();
-        let closure = async move {
-            loop {
-                let msgs = match network.recv_msgs(TransmitType::Direct).await {
-                    Ok(msgs) => Messages(msgs),
-                    Err(err) => {
-                        error!("failed to receive direct messages: {err}");
+        }
+    });
+    let network = net.clone();
+    let mut state = network_state.clone();
+    let direct_handle = async_spawn(async move {
+        loop {
+            let msgs = match network.recv_msgs(TransmitType::Direct).await {
+                Ok(msgs) => Messages(msgs),
+                Err(err) => {
+                    error!("failed to receive direct messages: {err}");
 
-                        // return zero messages so we sleep and try again
-                        Messages(vec![])
-                    }
-                };
-                if msgs.0.is_empty() {
-                    async_sleep(Duration::from_millis(100)).await;
-                } else {
-                    break msgs;
+                    // return zero messages so we sleep and try again
+                    Messages(vec![])
                 }
-            }
-        };
-        Some(boxed_sync(closure))
-    }));
-    let message_stream = Merge::new(broadcast_stream, direct_stream);
-    let network_state: NetworkMessageTaskState<_> = NetworkMessageTaskState {
-        event_stream: event_stream.clone(),
-    };
-    let registry = task_runner.registry.clone();
-    let network_message_handler = HandleMessage(Arc::new(
-        move |messages: either::Either<Messages<TYPES>, Messages<TYPES>>,
-              mut state: NetworkMessageTaskState<TYPES>| {
-            let messages = match messages {
-                either::Either::Left(messages) | either::Either::Right(messages) => messages,
             };
-            async move {
-                state.handle_messages(messages.0).await;
-                (None, state)
+            if msgs.0.is_empty() {
+                async_sleep(Duration::from_millis(100)).await;
```

Comment:
It's because the webserver uses a locked vector instead of a channel to communicating what it receives back to HotShot. So that implementation can, and often will return an empty vector.  I think we should keep as is and i'll create a task to clean this up when we get rid of the webserver.  For now though would probably break other things/tests and I think the sleep can't be hit for lip2p and memory network like you said.

Reply:
https://github.com/EspressoSystems/HotShot/issues/2558